File: internal/core/port/in/transfer.go
package in

import "momo-shell/internal/core/domain"

// ConflictPolicy resolves a name collision at the destination when
// uploading or downloading. The caller (frontend) diffs the destination
// listing up front and picks one policy per call -- the queue itself never
// prompts.
type ConflictPolicy string

const (
	ConflictOverwrite ConflictPolicy = "overwrite"
	ConflictRename    ConflictPolicy = "rename" // appends " (1)", " (2)", ...
	ConflictSkip      ConflictPolicy = "skip"
)

// TaskKind identifies what a transfer task is moving and in which direction.
type TaskKind string

const (
	TaskUpload   TaskKind = "upload"
	TaskDownload TaskKind = "download"
)

// TaskState is a transfer task's lifecycle state.
type TaskState string

const (
	TaskQueued   TaskState = "queued"
	TaskRunning  TaskState = "running"
	TaskDone     TaskState = "done"
	TaskFailed   TaskState = "failed"
	TaskCanceled TaskState = "canceled"
)

// TaskInfo is the read-only snapshot of one transfer task, published on
// transfer:task (every change) and returned by Tasks().
type TaskInfo struct {
	ID          string
	SessionID   string
	Kind        TaskKind
	State       TaskState
	Src         string
	Dst         string
	CurrentFile string
	Bytes       int64
	Total       int64
	Offset      int64 // reserved for v2 resume; always 0 in v1
	Error       string
}

// TransferUseCase is the driving port for SFTP file-browser and transfer
// operations. ZMODEM (rz/sz) is layered on top of this same interface
// starting in M5; M1-M3 only exercise the SFTP path.
type TransferUseCase interface {
	ListRemoteDir(sessionID, path string) ([]domain.RemoteEntry, error)
	HomeDir(sessionID string) (string, error)
	StatRemote(sessionID, path string) (entry domain.RemoteEntry, ok bool, err error)

	Upload(sessionID string, localPaths []string, remoteDir string, policy ConflictPolicy) (taskIDs []string, err error)
	Download(sessionID string, remotePaths []string, localDir string, policy ConflictPolicy) (taskIDs []string, err error)

	Mkdir(sessionID, path string) error
	Rename(sessionID, oldPath, newPath string) error
	Remove(sessionID, path string) error
	Chmod(sessionID, path string, mode uint32) error

	// CopyRemote stream-copies files (not directories) within one session's
	// remote file system into dstDir, keeping their basenames -- SFTP has no
	// server-side copy. Synchronous; files only in v1.
	CopyRemote(sessionID string, srcPaths []string, dstDir string) error

	Cancel(taskID string) error
	Tasks() []TaskInfo

	// StartZmodemSend answers a pending rz-upload detection (see
	// transfer:zmodem:{sessionId} "detected" events) with the files the
	// user picked or dropped.
	StartZmodemSend(sessionID string, localPaths []string) (taskID string, err error)
	// CancelZmodem aborts any in-flight ZMODEM transfer, or a pending
	// awaiting-send prompt, on a session.
	CancelZmodem(sessionID string) error
}

File: frontend/src/shared/api/transfer.ts
import {
  BrowseForDownloadDirectory,
  BrowseForUploadFiles,
  CancelTransfer,
  CancelZmodem,
  Chmod,
  CopyRemote,
  Download,
  HomeDir,
  ListRemoteDir,
  ListTasks,
  Mkdir,
  Remove,
  Rename,
  StartZmodemSend,
  StatRemote,
  Upload,
} from '../../../wailsjs/go/wails/TransferService'

export interface RemoteEntry {
  name: string
  path: string
  size: number
  mode: number
  modeText: string
  modTime: number // unix ms
  isDir: boolean
}

export type ConflictPolicy = 'overwrite' | 'rename' | 'skip'
export type TaskKind = 'upload' | 'download'
export type TaskState = 'queued' | 'running' | 'done' | 'failed' | 'canceled'

export interface TaskInfo {
  id: string
  sessionId: string
  kind: TaskKind
  state: TaskState
  src: string
  dst: string
  currentFile?: string
  bytes: number
  total: number
  offset: number
  error?: string
}

export async function listRemoteDir(sessionId: string, path: string): Promise<RemoteEntry[]> {
  return ListRemoteDir(sessionId, path)
}

export async function homeDir(sessionId: string): Promise<string> {
  return HomeDir(sessionId)
}

/** Wails only supports (value, error) returns; Go folds "not found" into a
 * nil pointer, which crosses the wire as null. */
export async function statRemote(sessionId: string, path: string): Promise<RemoteEntry | null> {
  return (await StatRemote(sessionId, path)) as unknown as RemoteEntry | null
}

/** Returns which of localPaths' basenames already exist in remoteDir --
 * callers prompt for an overwrite/rename/skip policy before uploadFiles
 * (the backend applies one policy per call, chosen up front by the caller;
 * it never prompts itself). */
export async function detectUploadConflicts(sessionId: string, localPaths: string[], remoteDir: string): Promise<string[]> {
  const entries = await listRemoteDir(sessionId, remoteDir)
  const existing = new Set(entries.map((e) => e.name))
  const basenames = localPaths.map((p) => p.split(/[\\/]/).pop() ?? p)
  return basenames.filter((name) => existing.has(name))
}

export async function uploadFiles(
  sessionId: string,
  localPaths: string[],
  remoteDir: string,
  policy: ConflictPolicy = 'overwrite'
): Promise<string[]> {
  return Upload(sessionId, localPaths, remoteDir, policy)
}

export async function downloadFiles(
  sessionId: string,
  remotePaths: string[],
  localDir: string,
  policy: ConflictPolicy = 'overwrite'
): Promise<string[]> {
  return Download(sessionId, remotePaths, localDir, policy)
}

export async function mkdirRemote(sessionId: string, path: string): Promise<void> {
  await Mkdir(sessionId, path)
}

export async function renameRemote(sessionId: string, oldPath: string, newPath: string): Promise<void> {
  await Rename(sessionId, oldPath, newPath)
}

export async function removeRemote(sessionId: string, path: string): Promise<void> {
  await Remove(sessionId, path)
}

export async function chmodRemote(sessionId: string, path: string, mode: number): Promise<void> {
  await Chmod(sessionId, path, mode)
}

/** Stream-copies files within sessionId's remote file system into dstDir --
 * SFTP has no server-side copy. Files only in v1. */
export async function copyRemoteFiles(sessionId: string, srcPaths: string[], dstDir: string): Promise<void> {
  await CopyRemote(sessionId, srcPaths, dstDir)
}

export async function cancelTransfer(taskId: string): Promise<void> {
  await CancelTransfer(taskId)
}

export async function listTasks(): Promise<TaskInfo[]> {
  return (await ListTasks()) as unknown as TaskInfo[]
}

/** Opens a native multi-file picker; returns [] if the user cancelled. */
export async function browseForUploadFiles(): Promise<string[]> {
  return BrowseForUploadFiles()
}

/** Opens a native folder picker; returns "" if the user cancelled. */
export async function browseForDownloadDirectory(): Promise<string> {
  return BrowseForDownloadDirectory()
}

/** Answers a pending transfer:zmodem "detected" (upload) event with the
 * files the user picked or dropped. Returns the new task's ID. */
export async function startZmodemSend(sessionId: string, localPaths: string[]): Promise<string> {
  return StartZmodemSend(sessionId, localPaths)
}

export async function cancelZmodem(sessionId: string): Promise<void> {
  await CancelZmodem(sessionId)
}
